content/blog/2022-05-03-e2e-cv-pt1.md

```diff
@@ -0,0 +1,316 @@
+which you can find
+[here](https://github.com/iterative/magnetic-tiles-defect/tree/main/notebooks).
+In short, there we have three notebooks:
+
+1. [`1_ProcessData.ipynb`](https://github.com/iterative/magnetic-tiles-defect/blob/main/notebooks/1_ProcessData.ipynb) -
+   downloads, processes, and organizes the data for easy loading into the
+   training process later
+2. [`2_TrainSegmentationModel.ipynb`](https://github.com/iterative/magnetic-tiles-defect/blob/main/notebooks/2_TrainSegmentationModel.ipynb) -
+   uses [`fastai`](https://github.com/fastai/fastai) Deep Learning framework to
+   train an image segmentation model
+3. [`3_Evaluate.ipynb`](https://github.com/iterative/magnetic-tiles-defect/blob/main/notebooks/3_Evaluate.ipynb) -
+   computes model performance on the test dataset
+
+Jupyter Notebook is by far the most popular tool for quick exploratory work when
+it comes to data analysis and modeling. However, it's not without
+[its own limitations](https://www.youtube.com/watch?v=7jiPeIFXb6U). One of the
+biggest issues of Jupyter is that it has no guardrails to ensure
+reproducibility, e.g. hidden states of variables and objects as well as the
+possibility to run cells out of order. While there are several projects that
+attempt to alleviate some of these issues (notably,
+[`nodebook`](https://github.com/stitchfix/nodebook),
+[`papermill`](https://github.com/nteract/papermill),
+[`nbdime`](https://github.com/jupyter/nbdime),
+[`nbval`](https://github.com/computationalmodelling/nbval),
+[`nbstripout`](https://github.com/kynan/nbstripout), and
+[`nbQA`](https://github.com/nbQA-dev/nbQA)), they don’t solve them completely.
+
+That's where the concepts of data versioning and ML pipelines come in.
+
+### Data Versioning
+
+Here, we'll briefly touch on DVC's functionality to version data and other
+artifacts (e.g. model weights).
+
+In most ML projects, training data changes gradually over time as new training
+instances (images in our case) get added while older ones might be removed.
+Simply creating snapshots of our training data at the time of training (e.g.
+labeling data directories with dates) quickly becomes unsustainable since these
+snapshots will contain many duplicates.
+
+A much better approach is to:
+
+1. track only the deltas between different versions of the datasets; and
+
+2. have the project’s git repository store only the reference links to the data
+   while the actual data is stored in a remote storage
+
+This is exactly what we can do with DVC by running only a couple of DVC commands
+and relying on DVC to handle all the underlying complexity:
+
+- compute hashes of files (images in our case) and move them to the cache
+- automatically manage reflinks to files stored in cache depending on the
+  version of the dataset being used
+- push and pull to different remote storage solutions using the same command
+  syntax
+- and, most importantly, integrate the process of data versioning with git-based
+  workflows
```

Comment:
IMO this is pretty deep and technical. Not sure if we need all of it. We could probably remove or find a better way to summarize at a higher level.

Reply:
paraphrased